### PR TITLE
Refactoring of type casters

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | 

Split `SimpleTypeCaster` to `PhpNativeTypeCaster` and `HydratorTypeCaster`.

### CODE REVIEW

File: src/TypeCaster/HydratorTypeCaster.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+use function is_array;
+
+/**
+ * Cast arrays to objects.
+ */
+final class HydratorTypeCaster implements TypeCasterInterface
+{
+    /**
+     * @param HydratorInterface $hydrator Hydrator to use to cast arrays to objects.
```
Comment (arogachev) [inline src/TypeCaster/HydratorTypeCaster.php:23]:
```suggestion
     * @param HydratorInterface $hydrator Hydrator to use for casting arrays to objects.
```

File: src/TypeCaster/HydratorTypeCaster.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        if ($type instanceof ReflectionNamedType) {
+            return $this->castInternal($value, $type);
+        }
+
+        if (!$type instanceof ReflectionUnionType) {
+            return Result::fail();
+        }
+
+        $types = array_filter(
```
Comment (arogachev) [inline src/TypeCaster/HydratorTypeCaster.php:44]: Maybe do it in one loop?

File: src/TypeCaster/HydratorTypeCaster.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+use ReflectionType;
+use ReflectionUnionType;
+use Yiisoft\Hydrator\HydratorInterface;
+use Yiisoft\Hydrator\Result;
+use Yiisoft\Hydrator\TypeCasterInterface;
+
+use function is_array;
+
+/**
+ * Cast arrays to objects.
```
Comment (arogachev) [inline src/TypeCaster/HydratorTypeCaster.php:18]:
```suggestion
 * Casts arrays to objects.
```

File: src/TypeCaster/HydratorTypeCaster.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ */
+final class HydratorTypeCaster implements TypeCasterInterface
```
Comment (arogachev) [inline src/TypeCaster/HydratorTypeCaster.php:20]: I think name could be a bit more specific.

**Review by arogachev — CHANGES_REQUESTED** [review]

File: src/TypeCaster/HydratorTypeCaster.php (hunk 4 above)
Comment (vjik) [reply src/TypeCaster/HydratorTypeCaster.php:20]: Do you have any suggestions?

File: src/TypeCaster/HydratorTypeCaster.php (hunk 2 above)
Comment (vjik) [reply src/TypeCaster/HydratorTypeCaster.php]: Done.

**Review by xepozz — APPROVED** [review]

File: src/TypeCaster/HydratorTypeCaster.php (hunk 4 above)
Comment (rustamwin) [reply src/TypeCaster/HydratorTypeCaster.php:20]: What about `ListTypeCaster` or `CollectionTypeCaster`?

**Review by rustamwin — APPROVED** [review]
Need tests.

**Comment by vjik** [comment]
> Need tests.

Yes. It well be done in separate PR.

**Pull request merged**